src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -43,9 +62,53 @@ class MetricsSource {
           metrics
         }
       }
-    )
+
+    SparkEnv.get.metricsSystem.registerSource(source)
+
+    source
+  }
+
+  def deregisterMetricSource(key: String, source: Source): Unit = {
+    metricsMap.remove(key)
+    SparkEnv.get.metricsSystem.removeSource(source)
   }
 }
 
 // Singleton to make sure each metric is only registered once.
 object MetricsSource extends MetricsSource
+
+class ExpiringCounter(key: String, source: Source) extends Counter {
+  private var (deathTimer, deathTimerTask): (Timer, TimerTask) = createTimer
```

Comment:
Unsure if the timer will survive long enough to deregister the metric upon task completion.
This will also not work on a crash of the executor, and the metrics will linger.

Another solution would be to move this to the driver, but the instance of the MetricSource (so that it can be removed from the MetricsSystem) gets a little more complicated. Looking into this approach now, though